Plugins/NativeCamera/Source/MediaStream.cpp

```diff
@@ -0,0 +1,335 @@
+        // ObjectWrap fullfills the MediaStream and MediaStreamTrack interfaces
+        auto tracks = Napi::Array::New(info.Env(), 1);
+        tracks.Set(uint32_t{0}, info.This());
+        return std::move(tracks);
+    }
+
+    Napi::Value MediaStream::GetAudioTracks(const Napi::CallbackInfo& info)
+    {
+        // Only video tracks are currently supported
+        return Napi::Array::New(info.Env(), 0);
+    }
+
+    Napi::Value MediaStream::ApplyConstraints(const Napi::CallbackInfo& info)
+    {
+        auto env = info.Env();
+        auto constraints = info[0].As<Napi::Object>();
+        
+        auto deferred{Napi::Promise::Deferred::New(env)};
+        auto promise{deferred.Promise()};
+
+        // Apply the constraints to the existing device
+        if(!UpdateConstraints(env, constraints))
+        {
+            // Setting the constraint to the capability failed
+            deferred.Reject(Napi::Error::New(env, std::runtime_error{"OverconstrainedError: Unable to match constraints to a supported camera configuration."}).Value());
+            return static_cast<Napi::Value>(promise);
+        }
+        deferred.Resolve(env.Undefined());
+
+        return static_cast<Napi::Value>(promise);
+    }
+
+    arcana::task<void, std::exception_ptr> MediaStream::ApplyInitialConstraints(Napi::Env env, Napi::Object constraints)
+    {
+        // A camera device hasn't been selected yet. Find the best device and open the stream
+        auto bestCamera = FindBestCameraStream(constraints);
+        m_cameraDevice = bestCamera.first;
+        m_cameraResolution = bestCamera.second;
+
+        // If m_cameraDevice is still null that means a camera device could not be found that meets the constraints
+        if (m_cameraDevice == nullptr)
+        {
+            // If no device could be fount to satisfy the constraints throw a "ConstraintError" to match the browser implementation
+            return arcana::task_from_error<void>(std::make_exception_ptr(std::runtime_error{"ConstraintError: Unable to match constraints to a supported camera configuration."}));
+        }
+
+        // Create a persistent ref to the constraints object so it isn't destructed during our async work
+        auto constraintsRef{ std::make_shared<Napi::ObjectReference>(Napi::Persistent(constraints)) };
+
+        return m_cameraImpl->Open(m_cameraDevice, m_cameraResolution).then(m_runtimeScheduler, arcana::cancellation::none(), [this, env, constraintsRef](const arcana::expected<Camera::Impl::CameraDimensions, std::exception_ptr>& result) {
+            if (result.has_error())
+            {
+                // Re-throw the error from opening the camera
+                throw result.error();
+            }
+            
+            auto cameraDimensions{result.value()};
+            this->Width = cameraDimensions.width;
+            this->Height = cameraDimensions.height;
+            
+            if(!UpdateConstraints(env, constraintsRef->Value()))
```

Comment:
```suggestion
            if (!UpdateConstraints(env, constraintsRef->Value()))
```

Reply:
How do we get into a scenario where we fail to apply constraints after the camera is open?  Shouldn't FindBestCameraStream guarantee we meet the constraints?